### PR TITLE
Handling of `@defer` in the query planner

### PR DESCRIPTION
**Pull request by pcmanus**
Experimental support for `@defer` in the query planner

### CODE REVIEW

**Review by benweatherman — APPROVED** [review]
🐐

File: internals-js/src/definitions.ts [hunk 1]
```diff
@@ -1072,8 +1099,50 @@ const graphQLBuiltInDirectivesSpecifications: readonly DirectiveSpecification[]
     locations: [DirectiveLocation.SCALAR],
     argumentFct: (schema) => ({ args: [{ name: 'url', type: new NonNullType(schema.stringType()) }], errors: [] })
   }),
+  // TODO: currently inconditionally adding @defer as the list of built-in. It's probably fine, but double check if we want to not do so when @defer-support is
```
Comment (glasser) [inline internals-js/src/definitions.ts:1102]: The general concept is that if `@defer` is "part of the schema" (visible via introspection if introspection is on, passes validation when included in an operation) then the server should "support defer" (though it could choose to not defer particular deferred fragments anyway).

File: internals-js/src/definitions.ts [hunk 2]
```diff
@@ -1072,8 +1099,50 @@ const graphQLBuiltInDirectivesSpecifications: readonly DirectiveSpecification[]
...
+  // not enabled or something (it would probably be hard to handle it at that point anyway but well...).
+  createDirectiveSpecification({
+    name: 'defer',
+    locations: [DirectiveLocation.FRAGMENT_SPREAD, DirectiveLocation.INLINE_FRAGMENT],
+    argumentFct: (schema) => ({
+      args: [
+        { name: 'label', type: schema.stringType() },
+        { name: 'if', type: schema.booleanType() },
```
Comment (glasser) [inline internals-js/src/definitions.ts:1110]: should be non-null (and with a default of true). same for `@stream`

File: internals-js/src/definitions.ts [hunk 3]
```diff
@@ -1072,8 +1099,50 @@ const graphQLBuiltInDirectivesSpecifications: readonly DirectiveSpecification[]
...
+        { name: 'initialCount', type: schema.intType(), defaultValue: 0 },
+        { name: 'if', type: schema.booleanType() },
+      ],
+      errors: [],
+    })
+  }),
 ];
 
+export type DeferDirectiveArgs = {
+  // TODO: we currently do not support variables for the defer label. Passing a label in a variable
```
Comment (glasser) [inline internals-js/src/definitions.ts:1133]: good news: this is also what [the working group decided](https://github.com/robrichard/defer-stream-wg/discussions/35) as shown in the spec PR in the "Defer And Stream Directive Labels Are Unique" validation.

File: query-planner-js/src/__tests__/buildPlan.defer.test.ts [hunk 4]
```diff
@@ -0,0 +1,3090 @@
+import { operationFromDocument, Schema, ServiceDefinition } from '@apollo/federation-internals';
+import gql from 'graphql-tag';
+import { QueryPlanner } from '@apollo/query-planner';
+import { composeAndCreatePlanner, composeAndCreatePlannerWithOptions } from "./buildPlan.test";
+
+function composeAndCreatePlannerWithDefer(...services: ServiceDefinition[]): [Schema, QueryPlanner] {
+  return composeAndCreatePlannerWithOptions(services, { deferStreamSupport: { enableDefer : true }});
+}
+
```
Comment (glasser) [inline query-planner-js/src/__tests__/buildPlan.defer.test.ts:9]: Great test suite.

One edge case I don't see tests for is when a single field is selected in the same selection set both deferred and not deferred (which in practice is likely to occur due to named fragment application) apparently the spec currently expects you to send the repeated field both in the initial response and a deferred response.

File: query-planner-js/src/config.ts [hunk 5]
```diff
@@ -15,4 +17,33 @@ export interface QueryPlannerConfig {
    * Defaults to true.
    */
   reuseQueryFragments?: boolean,
+
+  // Side-note: implemented as an object instead of single boolean because we expect to add more to this soon
+  // enough. In particular, once defer-passthrough to subgraphs is implemented, the idea would be to add a
+  // new `passthroughSubgraphs` option that is the list of subgraph to which we can pass-through some @defer
+  // (and it would be empty by default). Similarly, once we support @stream, grouping the options here will
+  // make sense too.
+  deferStreamSupport?: {
```
Comment (glasser) [inline query-planner-js/src/config.ts:26]: i'd suggest calling this `incrementalDelivery` or `incrementalDeliverySupport` (that's the more generic name for this feature less focused on the particular set of two directives which are currently envisioned)

File: query-planner-js/src/snapshotSerializers/queryPlanSerializer.ts [hunk 6]
```diff
@@ -62,21 +73,48 @@ function printNode(
...
+        ']' + config.spacingOuter +
+        indentation + '}';
+      break;
+    case 'Condition':
+      if (node.ifClause) {
+        if (node.elseClause) {
+          result +=
+            `Condition(if: \$${node.condition}) {` + config.spacingOuter +
+            indentationNext + printNode(node.ifClause, config, indentationNext, depth, refs, printer) + ',' + config.spacingOuter +
+            indentationNext + printNode(node.elseClause, config, indentationNext, depth, refs, printer) + config.spacingOuter +
```
Comment (glasser) [inline query-planner-js/src/snapshotSerializers/queryPlanSerializer.ts:103]: I'd find the tests with if/else easier to read if the serializer explicitly wrote "Else" or something in it

**Review by glasser — COMMENTED** [review]
I did a high level review of this (not going into the depths of the query planner, which I'm not familiar with) to see if I could bring anything helpful to the table and to help me understand the overall query plan structure for this defer implementation. Hope this helps!

**Comment by glasser** [comment]
Oh one other note. There are subtleties in the current spec around mutations. Normal GraphQL serial execution says you *completely* finish everything rooted at one root mutation field before moving on to the next one. (I hope Router does this correctly already? I actually don't see any tests of mutations in `buildPlan.test.ts` from before this PR.) However, the [decision for defer/stream](https://github.com/robrichard/defer-stream-wg/discussions/1) is that deferred blocks should be an exception here: if some stuff under a mutation root field is deferred, the next mutation root field can start.

I think your mutation tests cover this correctly but I'm not entirely sure I'm mentally parsing the plans correctly to verify :)

File: internals-js/src/definitions.ts (hunk 2 above)
Comment (glasser) [reply internals-js/src/definitions.ts:1110]: fixed in #2093, thanks

File: query-planner-js/src/__tests__/buildPlan.defer.test.ts (hunk 4 above)
Comment (glasser) [reply query-planner-js/src/__tests__/buildPlan.defer.test.ts:9]: fixed in #2093, thanks

File: query-planner-js/src/config.ts (hunk 5 above)
Comment (glasser) [reply query-planner-js/src/config.ts:26]: fixed in #2093, thanks

File: query-planner-js/src/snapshotSerializers/queryPlanSerializer.ts (hunk 6 above)
Comment (glasser) [reply query-planner-js/src/snapshotSerializers/queryPlanSerializer.ts:103]: fixed in #2093, thanks
